company-eudc.el

```diff
@@ -0,0 +1,155 @@
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+;;
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+;;    This library provides an interface to EUDC. EUDC is the Emacs
+;;    Unified Directory Client, a common interface to directory
+;;    servers and contact information, and part of core Emacs. EUDC
+;;    can consult mutiple servers for each query, providing the
+;;    combined results through a single interface. EUDC has backends
+;;    for LDAP servers, and to the BBDB.
+
+;;; Usage:
+;;    (require 'company-eudc)
+
+;;; Code:
+
+;;{{{      Internal cooking
+
+(require 'company)
+(require 'cl-lib)
+
+(defun company-eudc-email-address-from-alist (rec)
+  "Generate an email address string (\"first last <email>\") from
+  an attribute list."
+  (let* ((name (cdr (assoc 'name rec)))
+	 (firstname (cdr (assoc 'firstname rec)))
+	 (email (cdr (assoc 'email rec)))
+	 (candidate ""))
+    (if firstname
+	(setq candidate (format "%s %s <%s>" firstname name email))
+      (setq candidate (format "%s <%s>" name email)))
+    candidate))
+
+;;}}}
+
+;;{{{      High-level interfaces
+
+(defgroup company-eudc nil
+  "Completion backend for EUDC."
+  :group 'company)
+
+(defcustom company-eudc-modes '(message-mode notmuch-message-mode mail-mode)
+  "Major modes in which `company-eudc' may complete."
+  :type '(repeat (symbol :tag "Major mode")))
+
+;;;###autoload
+(defun company-eudc (command &optional arg &rest ignored)
+  "`company-mode' completion backend for EUDC.
+EUDC is the Emacs Unified Directory Client, a common interface to
+directory servers and contact information.
+
+Completion will be attempted in the To, Cc, Bcc, From, and
+Reply-To header fields of message buffers only. I.e. you can not
+use this backend for completing email addresses outside a message
+header.
+
+To load the `company-eudc' backend, simply do
+
+    (require 'company-eudc)
+
+With this alone, no completions from EUDC will be offered
+yet. You will additionally need to decide how you want to use
+it. Since EUDC can query remote resources (such as e.g. LDAP
+servers), the completion process may take a while. Depending on
+how you have configured `company-mode', this may result in Emacs
+being blocked for extensive periods of time. To accommodate this,
+the `company-eudc' backend can either install itself into the
+list of `company-mode' backends for `message-mode' (see
+`company-eudc-activate-autocomplete'), or provide autocompletion
+through `company-mode' only when bound to a key (see
+`company-eudc-expand-inline')."
+  (interactive (list 'interactive))
+  (pcase command
+    (`interactive (company-begin-backend 'company-eudc))
+    (`prefix (and (memq major-mode company-eudc-modes)
```

Comment:
Rather than using `memq` would say `derived-mode-p` be better? ie. just `(derived-mode-p message-mode)` so that say users of `mu4e` don't have to manually add say `mu4e-compose-mode` to `company-eudc-modes`?